File: V5_demo/wss_demo/sbe/bbo/sbe_bbo_integration.py
import json
import logging
import struct
import threading
import time
from datetime import datetime
from typing import Dict, Any

import websocket

logging.basicConfig(
    filename="logfile_wrapper.log",
    level=logging.INFO,
    format="%(asctime)s %(levelname)s %(message)s",
)

# Change symbol/topic as you wish
TOPIC = "ob.rpi.1.sbe.BTCUSDT"
WS_URL = "wss://stream-testnet.bybits.org/v5/realtime/spot"


class SBEBestOBRpiParser:
    """
    Parser for BestOBRpiEvent (template_id = 20000) per XML schema:

    ts(int64), seq(int64), cts(int64), u(int64),
    askNormalPrice(int64), askNormalSize(int64),
    askRpiPrice(int64), askRpiSize(int64),
    bidNormalPrice(int64), bidNormalSize(int64),
    bidRpiPrice(int64), bidRpiSize(int64),
    priceExponent(int8), sizeExponent(int8),
    symbol(varString8)

    All values are little-endian.
    """

    def __init__(self) -> None:
        # Header: blockLength, templateId, schemaId, version
        self.header_fmt = "<HHHH"
        self.header_sz = struct.calcsize(self.header_fmt)

        # 12 x int64 + 2 x int8:
        # ts, seq, cts, u,
        # askNormalPrice, askNormalSize, askRpiPrice, askRpiSize,
        # bidNormalPrice, bidNormalSize, bidRpiPrice, bidRpiSize,
        # priceExponent, sizeExponent
        self.body_fmt = "<" + ("q" * 12) + "bb"
        self.body_sz = struct.calcsize(self.body_fmt)

        self.target_template_id = 20000

    def _parse_header(self, data: bytes) -> Dict[str, Any]:
        if len(data) < self.header_sz:
            raise ValueError("insufficient data for SBE header")

        block_length, template_id, schema_id, version = struct.unpack_from(
            self.header_fmt, data, 0
        )

        return {
            "block_length": block_length,
            "template_id": template_id,
            "schema_id": schema_id,
            "version": version,
        }

    @staticmethod
    def _parse_varstring8(data: bytes, offset: int) -> tuple[str, int]:
        if offset + 1 > len(data):
            raise ValueError("insufficient data for varString8 length")

        (length,) = struct.unpack_from("<B", data, offset)
        offset += 1

        if offset + length > len(data):
            raise ValueError("insufficient data for varString8 bytes")

        s = data[offset : offset + length].decode("utf-8")
        offset += length
        return s, offset

    @staticmethod
    def _apply_exponent(value: int, exponent: int) -> float:
        # Exponent is for decimal point positioning.
        # If exponent = 2 and value=1060342500 -> 10603425.00
        return value / (10 ** exponent) if exponent >= 0 else value * (
            10 ** (-exponent)
        )

    def parse(self, data: bytes) -> Dict[str, Any]:
        hdr = self._parse_header(data)
        if hdr["template_id"] != self.target_template_id:
            raise NotImplementedError(
                f"unsupported template_id={hdr['template_id']}"
            )

        if len(data) < self.header_sz + self.body_sz:
            raise ValueError("insufficient data for BestOBRpiEvent body")

        fields = struct.unpack_from(self.body_fmt, data, self.header_sz)
        (
            ts,
            seq,
            cts,
            u,
            ask_np_m,
            ask_ns_m,
            ask_rp_m,
            ask_rs_m,
            bid_np_m,
            bid_ns_m,
            bid_rp_m,
            bid_rs_m,
            price_exp,
            size_exp,
        ) = fields

        offset = self.header_sz + self.body_sz
        symbol, offset = self._parse_varstring8(data, offset)

        # Apply exponents
        ask_np = self._apply_exponent(ask_np_m, price_exp)
        ask_ns = self._apply_exponent(ask_ns_m, size_exp)
        ask_rp = self._apply_exponent(ask_rp_m, price_exp)
        ask_rs = self._apply_exponent(ask_rs_m, size_exp)
        bid_np = self._apply_exponent(bid_np_m, price_exp)
        bid_ns = self._apply_exponent(bid_ns_m, size_exp)
        bid_rp = self._apply_exponent(bid_rp_m, price_exp)
        bid_rs = self._apply_exponent(bid_rs_m, size_exp)

        return {
            "header": hdr,
            "ts": ts,
            "seq": seq,
            "cts": cts,
            "u": u,
            "price_exponent": price_exp,
            "size_exponent": size_exp,
            "symbol": symbol,
            # Normal book (best)
            "ask_normal_price": ask_np,
            "ask_normal_size": ask_ns,
            "bid_normal_price": bid_np,
            "bid_normal_size": bid_ns,
            # RPI book (best)
            "ask_rpi_price": ask_rp,
            "ask_rpi_size": ask_rs,
            "bid_rpi_price": bid_rp,
            "bid_rpi_size": bid_rs,
            "parsed_length": offset,
        }


parser = SBEBestOBRpiParser()


# --------------------------- WebSocket handlers ---------------------------

def on_message(ws, message):
    try:
        # Binary SBE frames; text frames for control/acks/errors
        if isinstance(message, (bytes, bytearray)):
            decoded = parser.parse(message)
            logging.info(
                "SBE %s seq=%s u=%s "
                "NORM bid=%.8f@%.8f ask=%.8f@%.8f "
                "RPI bid=%.8f@%.8f ask=%.8f@%.8f ts=%s",
                decoded["symbol"],
                decoded["seq"],
                decoded["u"],
                decoded["bid_normal_price"],
                decoded["bid_normal_size"],
                decoded["ask_normal_price"],
                decoded["ask_normal_size"],
                decoded["bid_rpi_price"],
                decoded["bid_rpi_size"],
                decoded["ask_rpi_price"],
                decoded["ask_rpi_size"],
                decoded["ts"],
            )
            print(
                f"{decoded['symbol']} u={decoded['u']} "
                f"NORM: {decoded['bid_normal_price']:.8f} x {decoded['bid_normal_size']:.8f} "
                f"| {decoded['ask_normal_price']:.8f} x {decoded['ask_normal_size']:.8f} "
                f"RPI: {decoded['bid_rpi_price']:.8f} x {decoded['bid_rpi_size']:.8f} "
                f"| {decoded['ask_rpi_price']:.8f} x {decoded['ask_rpi_size']:.8f} "
                f"(seq={decoded['seq']} ts={decoded['ts']})"
            )
        else:
            try:
                obj = json.loads(message)
                logging.info("TEXT %s", obj)
                print(obj)
            except json.JSONDecodeError:
                logging.warning("non-JSON text frame: %r", message)
    except Exception as e:
        logging.exception("decode error: %s", e)
        print("decode error:", e)


def on_error(ws, error):
    print("WS error:", error)
    logging.error("WS error: %s", error)


def on_close(ws, *_):
    print("### connection closed ###")
    logging.info("connection closed")


def on_open(ws):
    print("opened")
    sub = {"op": "subscribe", "args": [TOPIC]}
    ws.send(json.dumps(sub))
    print("subscribed:", TOPIC)

    threading.Thread(target=ping_per, args=(ws,), daemon=True).start()
    threading.Thread(target=manage_subscription, args=(ws,), daemon=True).start()


def manage_subscription(ws):
    # demo: unsubscribe/resubscribe once
    time.sleep(20)
    ws.send(json.dumps({"op": "unsubscribe", "args": [TOPIC]}))
    print("unsubscribed:", TOPIC)
    time.sleep(5)
    ws.send(json.dumps({"op": "subscribe", "args": [TOPIC]}))
    print("resubscribed:", TOPIC)


def ping_per(ws):
    while True:
        try:
            ws.send(json.dumps({"op": "ping"}))
        except Exception:
            return
        time.sleep(10)


def on_pong(ws, *_):
    print("pong received")


def on_ping(ws, *_):
    print("ping received @", datetime.now().strftime("%Y-%m-%d %H:%M:%S"))


def connWS():
    ws = websocket.WebSocketApp(
        WS_URL,
        on_open=on_open,
        on_message=on_message,
        on_error=on_error,
        on_close=on_close,
        on_ping=on_ping,
        on_pong=on_pong,
    )
    ws.run_forever(ping_interval=20, ping_timeout=10)


if __name__ == "__main__":
    websocket.enableTrace(False)
    connWS()

File: V5_demo/wss_demo/sbe/bbo/sbe_bbo_integration.js
const WebSocket = require('ws');
const fs = require('fs');

// Change symbol/topic as you wish
const TOPIC = "ob.rpi.1.sbe.BTCUSDT";
const WS_URL = "wss://stream-testnet.bybits.org/v5/realtime/spot";

// -------------------------------------------------------------------
// Logging setup
// -------------------------------------------------------------------

function setupLogging() {
    const logStream = fs.createWriteStream('logfile_wrapper.log', { flags: 'a' });

    function log(level, message) {
        const timestamp = new Date().toISOString();
        const logMessage = `${timestamp} ${level} ${message}\n`;
        logStream.write(logMessage);
        console.log(`[${level}] ${message}`);
    }

    return {
        info: (msg) => log('INFO', msg),
        warning: (msg) => log('WARNING', msg),
        error: (msg) => log('ERROR', msg),
        exception: (error, context) => {
            const msg = `${context}: ${error.message}\n${error.stack}`;
            log('ERROR', msg);
        }
    };
}

const logger = setupLogging();

// -------------------------------------------------------------------
// SBE Parser for BestOBRpiEvent (template_id = 20000)
// -------------------------------------------------------------------

class SBEBestOBRpiParser {
    constructor() {
        // Header: blockLength, templateId, schemaId, version
        this.headerSz = 8; // 4 * uint16 = 8 bytes

        // 12 x int64 + 2 x int8
        this.bodySz = 12 * 8 + 2 * 1; // 98 bytes

        this.targetTemplateId = 20000;
    }

    _parseHeader(data, offset = 0) {
        if (data.length < offset + this.headerSz) {
            throw new Error("insufficient data for SBE header");
        }

        const blockLength = data.readUInt16LE(offset);
        const templateId = data.readUInt16LE(offset + 2);
        const schemaId = data.readUInt16LE(offset + 4);
        const version = data.readUInt16LE(offset + 6);

        return {
            blockLength,
            templateId,
            schemaId,
            version,
            offset: offset + this.headerSz
        };
    }

    _parseVarString8(data, offset) {
        if (offset + 1 > data.length) {
            throw new Error("insufficient data for varString8 length");
        }

        const length = data.readUInt8(offset);
        offset += 1;

        if (length === 0) {
            return { string: "", offset };
        }

        if (offset + length > data.length) {
            throw new Error("insufficient data for varString8 bytes");
        }

        const string = data.toString('utf8', offset, offset + length);
        offset += length;

        return { string, offset };
    }

    _applyExponent(value, exponent) {
        if (exponent >= 0) {
            return Number(value) / Math.pow(10, exponent);
        } else {
            return Number(value) * Math.pow(10, -exponent);
        }
    }

    parse(data) {
        const headerResult = this._parseHeader(data);
        if (headerResult.templateId !== this.targetTemplateId) {
            //throw new Error(`unsupported template_id=${headerResult.templateId}`);
            return null;
        }

        let offset = headerResult.offset;

        if (offset + this.bodySz > data.length) {
            throw new Error("insufficient data for BestOBRpiEvent body");
        }

        // Parse 12 int64 fields
        const ts = data.readBigInt64LE(offset);
        const seq = data.readBigInt64LE(offset + 8);
        const cts = data.readBigInt64LE(offset + 16);
        const u = data.readBigInt64LE(offset + 24);
        const askNpM = data.readBigInt64LE(offset + 32);
        const askNsM = data.readBigInt64LE(offset + 40);
        const askRpM = data.readBigInt64LE(offset + 48);
        const askRsM = data.readBigInt64LE(offset + 56);
        const bidNpM = data.readBigInt64LE(offset + 64);
        const bidNsM = data.readBigInt64LE(offset + 72);
        const bidRpM = data.readBigInt64LE(offset + 80);
        const bidRsM = data.readBigInt64LE(offset + 88);

        // Parse 2 int8 fields
        const priceExp = data.readInt8(offset + 96);
        const sizeExp = data.readInt8(offset + 97);

        offset += this.bodySz;

        // Parse symbol
        const symbolResult = this._parseVarString8(data, offset);
        const symbol = symbolResult.string;
        offset = symbolResult.offset;

        // Apply exponents
        const askNp = this._applyExponent(askNpM, priceExp);
        const askNs = this._applyExponent(askNsM, sizeExp);
        const askRp = this._applyExponent(askRpM, priceExp);
        const askRs = this._applyExponent(askRsM, sizeExp);
        const bidNp = this._applyExponent(bidNpM, priceExp);
        const bidNs = this._applyExponent(bidNsM, sizeExp);
        const bidRp = this._applyExponent(bidRpM, priceExp);
        const bidRs = this._applyExponent(bidRsM, sizeExp);

        return {
            header: {
                blockLength: headerResult.blockLength,
                templateId: headerResult.templateId,
                schemaId: headerResult.schemaId,
                version: headerResult.version
            },
            ts: ts.toString(),
            seq: seq.toString(),
            cts: cts.toString(),
            u: u.toString(),
            priceExponent: priceExp,
            sizeExponent: sizeExp,
            symbol: symbol,
            // Normal book (best)
            askNormalPrice: askNp,
            askNormalSize: askNs,
            bidNormalPrice: bidNp,
            bidNormalSize: bidNs,
            // RPI book (best)
            askRpiPrice: askRp,
            askRpiSize: askRs,
            bidRpiPrice: bidRp,
            bidRpiSize: bidRs,
            parsedLength: offset
        };
    }
}

const parser = new SBEBestOBRpiParser();

// -------------------------------------------------------------------
// WebSocket handlers
// -------------------------------------------------------------------

function onMessage(data) {
    try {
        if (Buffer.isBuffer(data)) {
            const decoded = parser.parse(data);
            if (decoded !== null && decoded !== undefined) {
                logger.info(
                    `SBE ${decoded.symbol} seq=${decoded.seq} u=${decoded.u} ` +
                    `NORM bid=${decoded.bidNormalPrice.toFixed(8)}@${decoded.bidNormalSize.toFixed(8)} ` +
                    `ask=${decoded.askNormalPrice.toFixed(8)}@${decoded.askNormalSize.toFixed(8)} ` +
                    `RPI bid=${decoded.bidRpiPrice.toFixed(8)}@${decoded.bidRpiSize.toFixed(8)} ` +
                    `ask=${decoded.askRpiPrice.toFixed(8)}@${decoded.askRpiSize.toFixed(8)} ts=${decoded.ts}`
                );

                console.log(
                    `${decoded.symbol} u=${decoded.u} ` +
                    `NORM: ${decoded.bidNormalPrice.toFixed(8)} x ${decoded.bidNormalSize.toFixed(8)} ` +
                    `| ${decoded.askNormalPrice.toFixed(8)} x ${decoded.askNormalSize.toFixed(8)} ` +
                    `RPI: ${decoded.bidRpiPrice.toFixed(8)} x ${decoded.bidRpiSize.toFixed(8)} ` +
                    `| ${decoded.askRpiPrice.toFixed(8)} x ${decoded.askRpiSize.toFixed(8)} ` +
                    `(seq=${decoded.seq} ts=${decoded.ts})`
                );
            }


        } else {
            try {
                const obj = JSON.parse(data);
                logger.info(`TEXT ${JSON.stringify(obj)}`);
                console.log(obj);
            } catch (jsonError) {
                logger.warning(`non-JSON text frame: ${data}`);
            }
        }
    } catch (error) {
        logger.exception(error, "decode error");
        console.log("decode error:", error.message);
    }
}

function onError(error) {
    console.log("WS error:", error);
    logger.error(`WS error: ${error.message}`);
}

function onClose() {
    console.log("### connection closed ###");
    logger.info("connection closed");
}

function pingPeriodically(ws) {
    setInterval(() => {
        try {
            if (ws.readyState === WebSocket.OPEN) {
                ws.send(JSON.stringify({ op: "ping" }));
            }
        } catch (error) {
            console.log("Ping error:", error.message);
        }
    }, 10000);
}

function manageSubscription(ws) {
    // demo: unsubscribe/resubscribe once
    setTimeout(() => {
        if (ws.readyState === WebSocket.OPEN) {
            ws.send(JSON.stringify({ op: "unsubscribe", args: [TOPIC] }));
            console.log("unsubscribed:", TOPIC);
        }
    }, 20000);

    setTimeout(() => {
        if (ws.readyState === WebSocket.OPEN) {
            ws.send(JSON.stringify({ op: "subscribe", args: [TOPIC] }));
            console.log("resubscribed:", TOPIC);
        }
    }, 25000);
}

function onOpen(ws) {
    console.log("opened");
    const sub = { op: "subscribe", args: [TOPIC] };
    ws.send(JSON.stringify(sub));
    console.log("subscribed:", TOPIC);

    // Start ping interval
    pingPeriodically(ws);

    // Start subscription management
    manageSubscription(ws);
}

function onPong() {
    console.log("pong received");
}

function onPing() {
    const now = new Date();
    console.log("ping received @", now.toISOString().replace('T', ' ').substring(0, 19));
}

function connWS() {
    const ws = new WebSocket(WS_URL);

    ws.on('open', () => onOpen(ws));
    ws.on('message', onMessage);
    ws.on('error', onError);
    ws.on('close', onClose);
    ws.on('pong', onPing);

    // Set up ping interval
    setInterval(() => {
        if (ws.readyState === WebSocket.OPEN) {
            ws.ping();
        }
    }, 20000);

    return ws;
}

// -------------------------------------------------------------------
// Main execution
// -------------------------------------------------------------------

if (require.main === module) {
    try {
        const ws = connWS();

        // Handle graceful shutdown
        process.on('SIGINT', () => {
            console.log('Shutting down...');
            ws.close();
            process.exit(0);
        });

    } catch (error) {
        console.error('Failed to start WebSocket connection:', error);
        process.exit(1);
    }
}
